# TP NoSQL

Prénom : Thomas 

Nom : Belaid 

Pour ce TP noté vous avez trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [1]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [8]:
import pymongo
import pandas as pd

URI = "mongodb+srv://ThomasB:EN5Ch9hChLJLoS61@clustertp1.xpndk.mongodb.net/?retryWrites=true&w=majority&appName=ClusterTP1"
client = pymongo.MongoClient(URI)
db = client.TP

print("Collections: ", db.list_collection_names())
print("movies: ", db.movies)
print("hello")


Collections:  ['restaurants', 'movies']
movies:  Collection(Database(MongoClient(host=['clustertp1-shard-00-02.xpndk.mongodb.net:27017', 'clustertp1-shard-00-01.xpndk.mongodb.net:27017', 'clustertp1-shard-00-00.xpndk.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', appname='ClusterTP1', authsource='admin', replicaset='atlas-3ljdyg-shard-0', tls=True), 'TP'), 'movies')
hello


1. Combien de films sont dans la base de données ?

In [153]:
count = db.movies.count_documents({})

print("Question 1 :","Le nombre de film dans la base de donnée est de ,",count,"films")


Question 1 : Le nombre de film dans la base de donnée est de , 23530 films


Le nombre de film dans la base de donnée est de , 23530 films


2. Quel est le nombre moyen de votes par pays ? Trier les pays par nombre moyen de votes décroissant.

In [154]:
pipeline = [
    {"$unwind":"$countries"},
    {"$group": {"_id": "$countries", "average_votes": {"$avg": "$imdb.votes"}}},
    {"$sort": {"average_votes": -1}}
]

result =  db.movies.aggregate(pipeline)

print("Question 2", pd.DataFrame(list(result)))

Question 2                       _id  average_votes
0                   Malta  210571.800000
1                 Bahamas  206061.000000
2             New Zealand   66589.269231
3    United Arab Emirates   53495.833333
4                     USA   35016.703663
..                    ...            ...
153                Kosovo     104.500000
154               Moldova     104.000000
155           North Korea      92.666667
156                Brunei      51.000000
157               Vanuatu       6.000000

[158 rows x 2 columns]


Le nombre moyen à malte est de 2105710,8

3. Donner la liste des différents genres de films.

In [155]:
 db.movies.distinct("genres")

['Action',
 'Adventure',
 'Animation',
 'Biography',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Family',
 'Fantasy',
 'Film-Noir',
 'History',
 'Horror',
 'Music',
 'Musical',
 'Mystery',
 'News',
 'Romance',
 'Sci-Fi',
 'Short',
 'Sport',
 'Talk-Show',
 'Thriller',
 'War',
 'Western']

4. Combien de films ont une note IMDb supérieure à 8 ?

In [156]:
count =  db.movies.count_documents(
    {"imdb.rating": {"$gt": 8}}
)

print("Question 4","Le nombre de film avec une note IMDb supérieure à 8 est de ", count)


Question 4 Le nombre de film avec une note IMDb supérieure à 8 est de  1177


 Le nombre de film avec une note IMDb supérieure à 8 est de 1117

5. Afficher le titre, l'année de sortie et les réalisateurs de tous les films (ne pas afficher l'ID).

In [159]:

result = {"_id": 0, "title": 1, "year": 1, "directors": 1}

data=pd.DataFrame(list(db.movies.find({}, result)))
data

title  \
0      Intolerance: Love's Struggle Throughout the Ages   
1                             The Poor Little Rich Girl   
2                                               The Kid   
3                                        Miss Lulu Bett   
4                                          Safety Last!   
...                                                 ...   
23525                                            Emelie   
23526                                    Techo y comida   
23527                                             Nahid   
23528                                  Coin Locker Girl   
23529                                 Tryapichnyy soyuz   

                            directors  year  
0                     [D.W. Griffith]  1916  
1                  [Maurice Tourneur]  1917  
2                   [Charles Chaplin]  1921  
3               [William C. de Mille]  1921  
4      [Fred C. Newmeyer, Sam Taylor]  1923  
...                               ...   ...  
23525                [Michael Thelin]  2015  
23526      [Juan Miguel del Castillo]  2015  
23527                [Ida Panahandeh]  2015  
23528                   [Jun-hee Han]  2015  
23529            [Mikhail Mestetskiy]  2015  

[23530 rows x 3 columns]

Le film Intolerance: Love's Struggle Throughout the Ages  a pour réalisateur D.W Griffith est sorti en 1916

6. Afficher le titre et le plot des films réalisés par D.W. Griffith.

In [161]:
result = {"_id": 0, "title": 1, "plot": 1}
data=pd.DataFrame(list(db.movies.find({"directors": "D.W. Griffith"}, result)))
data

plot  \
0  The story of a poor young woman, separated by ...   
1  A greedy tycoon decides, on a whim, to corner ...   
2  A frail waif, abused by her brutal boxer fathe...   
3  The Civil War divides friends and destroys fam...   

                                              title  
0  Intolerance: Love's Struggle Throughout the Ages  
1                                 A Corner in Wheat  
2    Broken Blossoms or The Yellow Man and the Girl  
3                             The Birth of a Nation

Le film avec le debut de plot "The story of a poor young woman, separated by" a pour titre  Intolerance: Love's Struggle Throughout the Ages   est a 
été réaliser par D.W. Griffith.

7. Quels sont les 5 acteurs les plus populaires (ceux qui ont joué dans le plus de films) ?

In [162]:
pipeline = [
    {"$unwind": "$cast"},
    {"$group": {"_id": "$cast", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}},
    {"$limit": 5}
]

result = pd.DataFrame(list(db.movies.aggregate(pipeline)))
result



_id  count
0      Gèrard Depardieu     68
1        Robert De Niro     60
2         Michael Caine     52
3  Marcello Mastroianni     50
4         Max von Sydow     49

Les 5 acteurs les plus populaires sont  Gèrard Depardieu  avec 68 films, Robert De Niro avec 60 films,  Michael Caine avec 52 films  Marcello Mastroianni avec 50 films,   Max von Sydow avec 49 films

8. Quelles sont les 6 années avec le plus de films produits ?

In [163]:
pipeline = [
    {"$group": {"_id": "$year", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}},
    {"$limit": 6}
]

result=pd.DataFrame(list(db.movies.aggregate(pipeline)))
result


_id  count
0  2013   1220
1  2014   1147
2  2012   1109
3  2011   1040
4  2009    990
5  2010    970

Les années les plus productionfs sont 2013 avec 1220 films, 2014 avec 1147 films, 2012 avec 1109 films, 2011 avec 1040 films, 2009 avec 990 films
2010

9. Combien de films ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 ?

In [164]:
pipeline = [
    {"$match": {"year": {"$lt": 2000}, "imdb.rating": {"$gt": 7}}},
    {"$count": "Nombre_film"}
]

result=list(db.movies.aggregate(pipeline))
result


[{'Nombre_film': 4747}]

Le nombre de film qui ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 est de 4747

10. Donner les 20 réalisateurs qui ont réalisé le plus de films (afficher seulement leur nom et le nombre de films).

In [165]:
pipeline = [
    {"$unwind": "$directors"},
    {"$group": {"_id": "$directors", "Comptage": {"$sum": 1}}},
    {"$sort": {"Comptage": -1}},
    {"$limit": 20}
]

result = pd.DataFrame(list(db.movies.aggregate(pipeline)))
result


_id  Comptage
0         Woody Allen        40
1           John Ford        35
2         John Huston        34
3       Takashi Miike        34
4       Werner Herzog        33
5     Martin Scorsese        32
6    Alfred Hitchcock        31
7        Sidney Lumet        30
8       Michael Apted        29
9        George Cukor        29
10   Steven Spielberg        29
11    Mario Monicelli        29
12      Robert Altman        28
13          Spike Lee        28
14  Steven Soderbergh        28
15     Clint Eastwood        27
16    Jean-Luc Godard        27
17        Wim Wenders        27
18         Johnnie To        27
19          Ken Loach        27

Le réalisateur qui réalisé le plus de film  est woody Allen avec 40 films

11. Quelle est la note IMDb moyenne des films sortis après 2010 ?

In [166]:
pipeline = [
    {"$match": {"year": {"$gt": 2010}}},
    {"$group": {"_id": None, "average_rating": {"$avg": "$imdb.rating"}}}
]

print(list(db.movies.aggregate(pipeline)))

[{'_id': None, 'average_rating': 6.544033171521035}]


la note IMDb moyenne des films sortis après 2010 est de 6.544033171521035

12. (bonus) Afficher les films dont le titre contient le mot "love".

In [117]:
pipeline = [
    {"$match": {"title": {"$regex": "^love", "$options": "i"}}},
    { "$project": { "_id": 0, "title": 1 } }]

print(pd.DataFrame(list(db.movies.aggregate(pipeline))))

                               title
0              Love Finds Andy Hardy
1                    Love Me Tonight
2                   Love on the Dole
3    Love Is a Many-Splendored Thing
4                Love Me or Leave Me
..                               ...
123              Love at First Fight
124                      Love Steaks
125        Love Is the Perfect Crime
126                        Lovemilla
127                             Love

[128 rows x 1 columns]


Il y a 128 films qui contiennent love donc :  Love Finds Andy Hardy,  Love Me Tonight et  Love on the Dole

13. (bonus) Afficher le titre des films ayant un score Rotten Tomatoes (viewer) supérieur à 75.


In [126]:
    
result = pd.DataFrame(list(db.movies.find({"tomatoes.viewer.meter": {"$gt": 75}},{ "_id": 0, "title": 1 ,"tomatoes.viewer.meter":1 } )))
result


title  \
0     Intolerance: Love's Struggle Throughout the Ages   
1                            The Poor Little Rich Girl   
2                                              The Kid   
3                                         Safety Last!   
4                                         Wild Oranges   
...                                                ...   
6597                                             Luton   
6598                                       Amira & Sam   
6599                                     Misconception   
6600                                             Grace   
6601                         Batman: Assault on Arkham   

                        tomatoes  
0      {'viewer': {'meter': 78}}  
1      {'viewer': {'meter': 77}}  
2      {'viewer': {'meter': 96}}  
3      {'viewer': {'meter': 93}}  
4      {'viewer': {'meter': 91}}  
...                          ...  
6597  {'viewer': {'meter': 100}}  
6598   {'viewer': {'meter': 76}}  
6599  {'viewer': {'meter': 100}}  
6600   {'viewer': {'meter': 86}}  
6601   {'viewer': {'meter': 86}}  

[6602 rows x 2 columns]

Un titre de film  ayant un score Rotten Tomatoes (viewer) supérieur à 75 est Intolerance: Love's Struggle Throughout the Ages avec 78 de score